### PR TITLE
Allow http transport to be used without wrapping Joomla or Koowa

### PR DESCRIPTION
**Pull request by johanjanssens**
Took another stab at trying to make the dispatcher transport more flexible and remove the need for tricks with headers. This PR does the following:

### 1. Dispatcher flush() action

Added a flush() action to the dispatcher() which handles flushing response to the output. By catching the flush action during a before.flush and returning false it's now  possible prevent the response from being flushed.

### 2. Moved Joomla application context handling to decoratable behavior

The new decoratable behavior handles setting up the Joomla application context during before.flush and will return false to prevent response from getting flushed output. This allows Joomla to continue handling the request.

The behavior also handles the decoration of the page, either using the joomla or koowa layout. This is handled during before.send and depends on the decorator defined in the view.

### 3. Removed tmpl=koowa and X-Flush-Response

Moved the responsibility of deciding how a view should be decorated into the view. This fits with the onion rendering approach, Koowa renders from inside to outside and it makes sense that the view can decide how it's being decorated. 

- Added a getDecorator() method to the view which returns the decorator() the view wants to be rendered in.
- Added a decorator() function to a template which returns the decorator set in the view and allows template helpers to understand what context they are used in.

This approach still allows for higher levels to handle the decoration otherwise if needed.

### Notes

Added a BC support for tmpl=koowa and for X-Flush-Response. If either is set the decorator will be forced to 'koowa.

### Reviews

Following is a list of code that might need to be reviewed as this change could break it.

https://github.com/joomlatools/textman/blob/b579d9a369659119bf629072c264b004b20bd580/code/plugins/system/textman/textman.php#L352

### CODE REVIEW

File: code/libraries/joomlatools/component/koowa/template/filter/meta.php [hunk 1]
```diff
@@ -26,7 +26,7 @@ public function filter(&$text)
     {
         $meta    = $this->_parseTags($text);
 
-        if($this->getTemplate()->getParameters()->layout == 'koowa') {
+        if($this->getTemplate()->decorator() == 'joomla') {
```
Comment (ercanozkaya) [inline code/libraries/joomlatools/component/koowa/template/filter/meta.php]: @johanjanssens This should check for 'koowa' no?

File: code/libraries/joomlatools/component/koowa/template/filter/script.php [hunk 2]
```diff
@@ -31,7 +31,7 @@ public function filter(&$text)
     {
         $scripts = $this->_parseTags($text);
 
-        if($this->getTemplate()->getParameters()->layout == 'koowa') {
+        if($this->getTemplate()->decorator() == 'joomla') {
```
Comment (ercanozkaya) [inline code/libraries/joomlatools/component/koowa/template/filter/script.php]: @johanjanssens Same as above. This should check for 'koowa' no?

File: code/libraries/joomlatools/component/koowa/template/filter/style.php [hunk 3]
```diff
@@ -31,7 +31,7 @@ public function filter(&$text)
     {
         $styles = $this->_parseTags($text);
 
-        if($this->getTemplate()->getParameters()->layout == 'koowa') {
+        if($this->getTemplate()->decorator() == 'joomla') {
```
Comment (ercanozkaya) [inline code/libraries/joomlatools/component/koowa/template/filter/style.php]: @johanjanssens Same as above. This should check for 'koowa' no?

File: code/libraries/joomlatools/component/koowa/template/filter/title.php [hunk 4]
```diff
@@ -26,7 +26,7 @@ public function filter(&$text)
     {
         $links   = $this->_parseTags($text);
 
-        if($this->getTemplate()->getParameters()->layout == 'koowa') {
+        if($this->getTemplate()->decorator() == 'joomla') {
```
Comment (ercanozkaya) [inline code/libraries/joomlatools/component/koowa/template/filter/title.php]: @johanjanssens Same as above. This should check for 'koowa' no?

File: code/libraries/joomlatools/component/koowa/template/filter/meta.php (hunk 1 above)
Comment (johanjanssens) [reply code/libraries/joomlatools/component/koowa/template/filter/meta.php]: True, fixed.

File: code/libraries/joomlatools/component/koowa/template/filter/script.php (hunk 2 above)
Comment (johanjanssens) [reply code/libraries/joomlatools/component/koowa/template/filter/script.php]: True, fixed.

File: code/libraries/joomlatools/component/koowa/template/filter/style.php (hunk 3 above)
Comment (johanjanssens) [reply code/libraries/joomlatools/component/koowa/template/filter/style.php]: True, fixed.

File: code/libraries/joomlatools/component/koowa/template/filter/title.php (hunk 4 above)
Comment (johanjanssens) [reply code/libraries/joomlatools/component/koowa/template/filter/title.php]: True, fixed.

**Comment by johanjanssens** [comment]
@ercanozkaya About your comment, the decorator is added to ComKoowaViewHtml so any view that extends from this will have the decorator set, this is the default Koowa view. We could move the approach lower if you are using framework views?

**Comment by ercanozkaya** [comment]
@johanjanssens No I still extend from here so I have the method but the problem is its value. 

If you are asking page view what the layout is it'll correctly return joomla or koowa. However if you are in a docman view for example and ask the decorator it will return null since it never received the value in the first place right?

**Comment by johanjanssens** [comment]
I have set the default to 'joomla', which means that if you need to render in a 'koowa' context you need to change it in your view class. This is the same approach used before with the header or with tmpl=koowa. Only now it's driven by the view not the request. Am I missing something?

**Comment by johanjanssens** [comment]
@ercanozkaya Can you give this a final swirl. I also fixed the error handling, if you force an error and have debug on you should see our koowa debugger kick in.

**Comment by ercanozkaya** [comment]
@johanjanssens Fixed a final statement error. This is good to go 👍

File: code/libraries/joomlatools/component/koowa/template/filter/title.php (hunk 4 above)
Comment (harmis-tech) [reply code/libraries/joomlatools/component/koowa/template/filter/title.php]: Call to undefined method :decorator - this type of error occur when i open my component in joomla

**Pull request merged**